packages/client/src/components/UserContext.tsx

```diff
@@ -14,21 +12,30 @@ export const UserContext = createContext<{
 const persistedUsername = getStoredUsername();
 export const UserProvider: FC = ({ children }) => {
   const [username, setUsername] = useState<string>(persistedUsername ?? '');
-  if (!username) {
-    gameIO.emit(UserEvent.UsernameMissing);
-  }
+  const hasUserName = useRef(username !== '');
+
+  const handleUsernameCreated = (randomUsername: string) => {
+    setUsername(randomUsername);
+    persistUsername(randomUsername);
+    hasUserName.current = false;
+  };
 
   useEffect(() => {
     gameIO.on('connect', () => {
       console.log(`connect ${gameIO.id}`);
     });
-    const handleUsernameCreated = (randomUsername: string) => {
-      setUsername(randomUsername);
-      persistUsername(randomUsername);
-    };
-    gameIO.on(UserEvent.UsernameCreated, handleUsernameCreated);
+
+    // We check if username exists to avoid double-registering the callback.
+    // Same for de-registering.
+    if (!hasUserName.current) {
```

Comment:
maybe we can do the opposite and call this missingUsername? (btw name lowercase?)
It's just to avoid the double negation in the ifs.

Also does bad things happen if you do off about an event you didn't set? Otherwise I would skip the second check but anyway if it's needed good as it its

Reply:
I had it like `hasMissingUsername`/`isMissingUsername`, but I usually prefer naming variables positively, instead of thinking oppositely. Happy to change this one, since it's only used with `!`. What do you mean by _double regation in the ifs_? What is the _ifs_?

Reply:
I meant 2 if :-) IFs